File: kits/js/koth/game_objects.js
/**
 * All game objects for access by user
 */
const {
  Position
} = require("./map");
const GAME_CONSTANTS = require("./game_constants");

/**
 * holds all data related to a player
 */
class Player {
  constructor(teamid) {
    this.team = teamid;
    this.points = 0;
    /** array of all units */
    this.units = [];
    /** arrrary of all bases */
    this.bases = [];
  }
}

// all data related to a Base
class Base {
  constructor(teamid, x, y) {
    this.team = teamid;
    this.pos = new Position(x, y);
  }
  /** return the command to move unit in the given direction */
  spawnUnit() {
    return `c ${this.pos.x} ${this.pos.y}`;
  }
}

class Unit {
  constructor(teamid, unitid, x, y, lastRepairTurn, turn) {
    this.pos = new Position(x, y);
    this.team = teamid;
    this.id = unitid;
    this.lastRepairTurn = lastRepairTurn;
    this.matchTurn = turn;
  }
 
  /** return the command to move unit in the given direction */
  move(dir) {
    return `m ${this.id} ${dir}`;
  }

  getBreakdownLevel() {
    return Math.floor((this.matchTurn - this.lastRepairTurn) / GAME_CONSTANTS.PARAMETERS.BREAKDOWN_TURNS);
  }
}

module.exports = {
  Player,
  Base,
  Unit,
}

File: kits/python/energium/game_objects.py
from typing import List
from .position import Position
from .game_constants import GAME_CONSTANTS
import math
import sys
class Base:
    def __init__(self, team: int, x, y):
        self.team = team
        self.pos = Position(x, y)

    def spawn_unit(self):
        return 'c {} {}'.format(self.pos.x, self.pos.y)


class Unit:
    def __init__(self, team: int, unitid: int, x, y, last_repair_turn, turn):
        self.team = team
        self.id = unitid
        self.pos = Position(x, y)
        self.last_repair_turn = last_repair_turn
        self.match_turn = turn
    
    def get_breakdown_level(self):
        """
        returns the breakdown level of this unit
        """
        return (self.match_turn - self.last_repair_turn) / GAME_CONSTANTS['PARAMETERS']['BREAKDOWN_TURNS'];
    def move(self, dir):
        self.pos = self.pos.translate(dir, 1)
        return 'm {} {}'.format(self.id, dir)

class Player:
    energium: int
    team: int
    units: List[Unit]
    bases: List[Base]
    def __init__(self, team):
        self.team = team
        self.bases = []
        self.units = []
        self.energium = 0
